### PR TITLE
[lang] Fix merge and add mergeRecursive

### PR DESCRIPTION
**Pull request by cristianoc72**
Add `ArrayObject::mergeRecursive`  method and relative test.
Add a test to prove the bug in `ArrayObject::merge` method and fix it.

### CODE REVIEW

**Review by gossi — COMMENTED** [review]
Thanks for adding another missing test. Yet I'm "worried" you changed the signature of the `merge()` methods.

Before: `$arr->merge(1, 2, 3);`
After: `$arr->merge([1, 2, 3]);`

Whereas, before this was also possible: `$arr->merge(1, ['a', 'b']. 3);` (which is now a two-dimensional array.

Maybe I don't see the reason? Can you enlighten us?

**Comment by cristianoc72** [comment]
When I use a `merge` method, I expect to merge two or more arrays together. So, imho, the parameters of the method should be arrays (currently, it's a list of the elements of one array) e.g.:
```php
<?php

$obj = new ArrayObject($initialArray);

//Each array can be multidimensional
$arr1 = [.....];
$arr2 = [....];
$arr3 = [.....];

$obj->merge($arr1, $arr2, $arr3);
``` 
With this signature, I obtain a true merge: the elements of *$arr1*, *$arr2* and *$arr3* are included in *$initialArray* (see the test).
With the original signature, instead, the 3 arrays are added with a new numeric key for each:
```php
$initialArray = [
    //elements of the initial array
    .......
    [0] => [
                //elements of $arr1
                ........
            ],
    [1] => [
                //elements of $arr2
                .........
            ],
    [2] => [
                //elements of $arr3
                .......
            ]
];
```
and thinking about the original implementation it's correct, because we've merged the `func_get_args()` array.

But it's not a merge, it's an addiction: imho, the correct instruction to obtain this array should be `$obj->add($arr1, $arr2, $arr3)`.
The current implementation of the `add` method doesn't support the possibility of multiple addictions, but we can work on it.

**Comment by gossi** [comment]
Basically, your merge is now an `addAll()` and not what the original [`array_merge()`](https://www.php.net/array_merge) is.

For the collections there was `add()` and `addAll()`, with the distinct use-case of `add()` to put in exactly _one_ item and `addAll()` for multiple. Actually, we can do more clever here I think. What about:

- `add(...$items)` - add one to many
- `merge(...$arrays)` - flattens them in

Basically, we can get rid of those `addAll()` methods and just use one method that accepts spread parameters.

What do you think?

**Comment by cristianoc72** [comment]
> Basically, your merge is now an `addAll()` and not what the original [`array_merge()`](https://www.php.net/array_merge) is.

I do not agree: my function behaves exactly as `array_merge()` and I added a new test to prove it.
In `tests/lang/ArrayTest::testMergeMultidimensional()` if you manually write:
```php
<?php

$result = array_merge($cartoons, $addMarvel, $goNagai);
```
the `$result` ( `$expected` array, into the test ) is exactly the result of my merge function, while the original returns:
```php
<?php

$result = [
	'Marvel' => [
		'avengers' => [
			"Iron Man",
			"Hulk",
			"Captain America"
			]
		],
	'Dc' => [
		"Superman",
		"Bat Man"
		],
	[
		'Marvel' => [
			'X-Men' => [
				"Wolverine",
				"Professor X",
				"Magneto"
				]
			]
		],
	[
		'Go Nagai' => [
			"Mazinger-Z",
			"Great Mazinger",
			"Goldrake"
			]
		]
	];
```

**Comment by cristianoc72** [comment]
> For the collections there was `add()` and `addAll()`, with the distinct use-case of `add()` to put in exactly _one_ item and `addAll()` for multiple. Actually, we can do more clever here I think. What about:
> 
>     * `add(...$items)` - add one to many
> 
>     * `merge(...$arrays)` - flattens them in
> 
> 
> Basically, we can get rid of those `addAll()` methods and just use one method that accepts spread parameters.
> 
> What do you think?

Great idea! I'll work on it asap.

Firstly I'ld like to review the `ArrayObject` consistency because there are some methods, working perfectly with arrays, that seem to not take in consideration the associative arrays (i.e. `add` method seems to deal with only numeric indexes). My idea is to add some specific tests and see what happens.

File: src/lang/ArrayObject.php [hunk 1]
```diff
@@ -164,12 +164,29 @@ public function slice(int $offset, ?int $length = null, bool $preserveKeys = fal
...
-	 * @param mixed ... Variable list of arrays to merge.
+	 * @param mixed ...$toMerges Variable list of arrays to merge.
 	 *
 	 * @return ArrayObject $this
 	 */
-	public function merge(): self {
-		$this->array = array_merge($this->array, func_get_args());
+	public function merge(...$toMerges): self {
+		foreach ($toMerges as $toMerge) {
+			$this->array = array_merge($this->array, $toMerge);
```
Comment (gossi) [inline src/lang/ArrayObject.php]: would `array_merge($this->array, ...$toMerges);` work here?

File: tests/lang/ArrayTest.php [hunk 2]
```diff
@@ -398,4 +398,91 @@ public function testGetWithNotExistentIndex(): void {
...
+				'Bat Man'
+				],
+			'Go Nagai' => [
+				'Mazinger-Z',
+				'Great Mazinger',
+				'Goldrake'
+				]
+			];
+
+		$this->assertEquals($expected, $cartoons->merge($addMarvel, $goNagai)->toArray());
```
Comment (gossi) [inline tests/lang/ArrayTest.php]: I think the `$expected` then should not be handmade but also the result of `array_merge()` to stay in parity with the original function.

**Review by gossi — COMMENTED** [review]
Ah, nice. I may also be wrong about the `array_merge()` of php itself, thanks for insisting on it 😊

I left some comments, hope they are helpful.

File: src/lang/ArrayObject.php (hunk 1 above)
Comment (cristianoc72) [reply src/lang/ArrayObject.php]: Awesome catch!! I didn't understand that variadics could be used in such way. Thank you!

Fixed.

File: tests/lang/ArrayTest.php (hunk 2 above)
Comment (cristianoc72) [reply tests/lang/ArrayTest.php]: Agree

**Comment by cristianoc72** [comment]
@gossi can we merge this PR or Should I add the refactor of `add` methods, too? It could be better to submit a new PR, imho.

**Review by gossi — APPROVED** [review]

**Comment by gossi** [comment]
yap, looking good, although a little education for me was needed =)

**Comment by gossi** [comment]
Ah, I was about to open a PR for `add(...$items)` but I forgot, that the signature for that method is already: `add($element, ?int $index = null)` which takes two parameters. So it's good to keep it and have separate methods for `add()` and `addAll()`.

**Comment by cristianoc72** [comment]
> Ah, I was about to open a PR for `add(...$items)` but I forgot, that the signature for that method is already: `add($element, ?int $index = null)` which takes two parameters. So it's good to keep it and have separate methods for `add()` and `addAll()`.

A new PR is on the road: get a look if you like the idea.
